File: android/daemon/src/main/java/com/joyliveex/vjc/daemon/Crypto.java
package com.joyliveex.vjc.daemon;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * Session crypto — deterministic manual PBKDF2-HMAC-SHA256 + AES/GCM, identical to
 * VjcCrypto.cs (desktop) and VjcCrypto.kt (companion app). Manual PBKDF2 avoids
 * provider differences on older ART builds.
 */
public final class Crypto {
    private Crypto() {}

    public static byte[] pbkdf2Sha256(String password, byte[] salt, int iterations, int dkLen) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(password.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            byte[] dk = new byte[dkLen];
            int offset = 0;
            int block = 1;
            byte[] input = new byte[salt.length + 4];
            System.arraycopy(salt, 0, input, 0, salt.length);
            byte[] u = new byte[32];
            byte[] t = new byte[32];
            while (offset < dkLen) {
                int be = block;
                input[input.length - 4] = (byte) (be >>> 24);
                input[input.length - 3] = (byte) (be >>> 16);
                input[input.length - 2] = (byte) (be >>> 8);
                input[input.length - 1] = (byte) be;
                mac.reset();
                u = mac.doFinal(input);
                System.arraycopy(u, 0, t, 0, 32);
                for (int i = 1; i < iterations; i++) {
                    mac.reset();
                    u = mac.doFinal(u);
                    for (int j = 0; j < 32; j++) t[j] ^= u[j];
                }
                int take = Math.min(32, dkLen - offset);
                System.arraycopy(t, 0, dk, offset, take);
                offset += take;
                block++;
            }
            return dk;
        } catch (Exception e) {
            throw new RuntimeException("pbkdf2 failed", e);
        }
    }

    public static String normalizeCode(String code) {
        StringBuilder sb = new StringBuilder();
        for (char c : code.toUpperCase().toCharArray())
            if (Character.isLetterOrDigit(c)) sb.append(c);
        return sb.toString();
    }

    public static byte[] deriveSessionKey(String pairingCode, byte[] clientSalt, byte[] serverSalt) {
        byte[] salt = new byte[16];
        System.arraycopy(clientSalt, 0, salt, 0, 8);
        System.arraycopy(serverSalt, 0, salt, 8, 8);
        return pbkdf2Sha256(normalizeCode(pairingCode), salt, Constants.PBKDF2_ITERATIONS, 32);
    }

    public static int computeSessionId(byte[] clientSalt, byte[] serverSalt) {
        try {
            MessageDigest sha = MessageDigest.getInstance("SHA-256");
            byte[] h = sha.digest("VJC-session|".getBytes(StandardCharsets.US_ASCII));
            sha.reset();
            byte[] ctx = new byte[16];
            System.arraycopy(clientSalt, 0, ctx, 0, 8);
            System.arraycopy(serverSalt, 0, ctx, 8, 8);
            sha.update(h);
            byte[] full = sha.digest(ctx);
            return ((full[0] & 0xFF) << 24) | ((full[1] & 0xFF) << 16) | ((full[2] & 0xFF) << 8) | (full[3] & 0xFF);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    static void buildNonce(byte[] nonce, int direction, int sessionId, int seq) {
        nonce[0] = (byte) direction;
        nonce[1] = (byte) sessionId;
        nonce[2] = (byte) (sessionId >>> 8);
        nonce[3] = (byte) (sessionId >>> 16);
        nonce[4] = (byte) (sessionId >>> 24);
        nonce[5] = (byte) seq;
        nonce[6] = (byte) (seq >>> 8);
        nonce[7] = (byte) (seq >>> 16);
        nonce[8] = (byte) (seq >>> 24);
        nonce[9] = 0;
        nonce[10] = 0;
        nonce[11] = 0;
    }

    public static byte[] encrypt(byte[] key, int direction, int sessionId, int seq, byte[] plaintext) {
        try {
            byte[] nonce = new byte[12];
            buildNonce(nonce, direction, sessionId, seq);
            Cipher c = Cipher.getInstance("AES/GCM/NoPadding");
            c.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key, "AES"), new GCMParameterSpec(128, nonce));
            return c.doFinal(plaintext); // ct||tag
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static byte[] decrypt(byte[] key, int direction, int sessionId, int seq, byte[] sealed) {
        try {
            byte[] nonce = new byte[12];
            buildNonce(nonce, direction, sessionId, seq);
            Cipher c = Cipher.getInstance("AES/GCM/NoPadding");
            c.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key, "AES"), new GCMParameterSpec(128, nonce));
            return c.doFinal(sealed);
        } catch (Exception e) {
            return null;
        }
    }
}

File: android/app/src/main/java/com/joyliveex/vjc/network/Crypto.kt
package com.joyliveex.vjc.network

import java.security.MessageDigest
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * PBKDF2-HMAC-SHA256 (manual, RFC 8018) + AES-GCM — byte-identical to VjcCrypto.cs
 * and the daemon's Crypto.java.
 */
object VjcCrypto {

    fun pbkdf2(password: String, salt: ByteArray, iterations: Int, dkLen: Int): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(password.toByteArray(Charsets.UTF_8), "HmacSHA256"))
        val dk = ByteArray(dkLen)
        var offset = 0
        var block = 1
        val input = ByteArray(salt.size + 4)
        salt.copyInto(input)
        while (offset < dkLen) {
            input[input.size - 4] = (block ushr 24).toByte()
            input[input.size - 3] = (block ushr 16).toByte()
            input[input.size - 2] = (block ushr 8).toByte()
            input[input.size - 1] = block.toByte()
            var u = mac.doFinal(input)
            val t = u.copyOf()
            for (i in 1 until iterations) {
                u = mac.doFinal(u)
                for (j in 0 until 32) t[j] = (t[j].toInt() xor u[j].toInt()).toByte()
            }
            val take = minOf(32, dkLen - offset)
            t.copyInto(dk, offset, 0, take)
            offset += take
            block++
        }
        return dk
    }

    fun normalizeCode(code: String): String =
        code.uppercase().filter { it.isLetterOrDigit() }

    fun deriveSessionKey(code: String, clientSalt: ByteArray, serverSalt: ByteArray): ByteArray {
        val salt = ByteArray(16)
        clientSalt.copyInto(salt, 0, 0, 8)
        serverSalt.copyInto(salt, 8, 0, 8)
        return pbkdf2(normalizeCode(code), salt, Wire.PBKDF2_ITER, 32)
    }

    fun computeSessionId(clientSalt: ByteArray, serverSalt: ByteArray): Int {
        val sha = MessageDigest.getInstance("SHA-256")
        val h = sha.digest("VJC-session|".toByteArray(Charsets.US_ASCII))
        val ctx = ByteArray(16)
        clientSalt.copyInto(ctx, 0, 0, 8)
        serverSalt.copyInto(ctx, 8, 0, 8)
        val full = sha.digest(h + ctx)
        return ((full[0].toInt() and 0xFF) shl 24) or
            ((full[1].toInt() and 0xFF) shl 16) or
            ((full[2].toInt() and 0xFF) shl 8) or
            (full[3].toInt() and 0xFF)
    }

    private fun nonce(dir: Int, session: Int, seq: Int): ByteArray {
        val n = ByteArray(12)
        n[0] = dir.toByte()
        n[1] = session.toByte()
        n[2] = (session ushr 8).toByte()
        n[3] = (session ushr 16).toByte()
        n[4] = (session ushr 24).toByte()
        n[5] = seq.toByte()
        n[6] = (seq ushr 8).toByte()
        n[7] = (seq ushr 16).toByte()
        n[8] = (seq ushr 24).toByte()
        return n
    }

    fun encrypt(key: ByteArray, dir: Int, session: Int, seq: Int, plain: ByteArray): ByteArray {
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce(dir, session, seq)))
        return c.doFinal(plain)
    }

    fun decrypt(key: ByteArray, dir: Int, session: Int, seq: Int, sealed: ByteArray): ByteArray? {
        return try {
            val c = Cipher.getInstance("AES/GCM/NoPadding")
            c.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce(dir, session, seq)))
            c.doFinal(sealed)
        } catch (e: Exception) {
            null
        }
    }
}

object PairingCode {
    fun toWire(display: String): String = VjcCrypto.normalizeCode(display)
}
